### PR TITLE
Make the heartbeat writer use 2 pools

### LINKED ISSUE #9010: VTTablet heartbeat write should use DBA connection
#### Feature Description
When vttablet creates a heartbeatWriter, it takes connection parameters for both vt_app and vt_dba. However, when execute CREATE TABLE statement, it always uses vt_app: https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletserver/repltracker/writer.go#L173

We should use vt_dba for the `CREATE TABLE` statement, and vt_app user for the actual insert/update

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

Currently the heartbeat writer uses `vt_app` user to create the database, table and also run the upsert commands. This isn't correct since `vt_app` might not have the `CREATE` privilege, causing the create database and table commands to fail. This PR addresses this issue by creating 2 app pools, one for `vt_app` to run the insert commands and `vt_allprivs` to run the create commands.

As part of this PR, `WithDDL` has also changed, and now takes two functions instead of one in its `Execute` method. The second function, if provided, is used for running the DDL commands. This allows the caller of this function to have two different execution functions for DDLs and other queries. This change will be helpful in the future too, in all the other usages of `WithDDL` which are using the `vt_app` user for running the DDLs too.

## Related Issue(s)

- Fixes #9010 
- Implements the discussion https://github.com/vitessio/vitess/issues/9290#issuecomment-982127317

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/withddl/withddl.go [hunk 1]
```diff
@@ -52,17 +52,29 @@ func (wd *WithDDL) DDLs() []string {
...
 	}
-	qr, err := exec(query)
+	execDDL := execQuery
+	if fDDL != nil {
+		execDDL, err = wd.unify(ctx, fQuery)
+		if err != nil {
+			return nil, err
+		}
+	}
+	qr, err := execQuery(query)
```
Comment (harshit-gangal) [inline go/vt/withddl/withddl.go:77]: looks like here fDDL needs to be used inside.
Also, can you add a unit test that passes both functions?

**Review by sonne5 — COMMENTED** [review]
Thanks for working on this!

I wonder if it makes sense to do a round of auditing to make sure the create table ddl for the internal tables (e.g., heartbeat, schema migration, shard split, etc) are all using `vt_allprivs` instead of `vt_app`? We ran into a few other privilege cases where we end up switching vt_app to vt_dba for schema migration as well.

Related: idk if it would be an overkill, if we decide `vt_allprivs` will be responsible for creating all the internal meta tables, would it be helpful to add a check and raise errors if user end up not granting `CREATE` privilege to it?

**Comment by GuptaManan100** [comment]
@sonne5 you are right in that we need to go over and look at all the uses of `vt_app`. There are other places as well where we use vt_app for create statements. We will fix it too but it's going to be a larger refactor

File: go/vt/withddl/withddl.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/withddl/withddl.go]: Yes, fixed both now

**Comment by deepthi** [comment]
@sonne5 right now a whole bunch of things [fail](https://github.com/vitessio/vitess/pull/9313) if I remove `CREATE` from `vt_app`, so it's clear that we are relying on that. Also Manan and I went through the code and `vt_allprivs` is being used in only one place.
There is already a tracking [issue](https://github.com/vitessio/vitess/issues/6474) for this. We just need to do what that issue is saying.

**Review by harshit-gangal — APPROVED** [review]

**Review by deepthi — APPROVED** [review]
Nice work!

**Review by sougou — APPROVED** [review]

**Pull request merged**